aodndata/aatams_sattag_qc_dm/csv_schema.py

```diff
@@ -0,0 +1,279 @@
+          SchemaError: if any key pair is different.
+
+        """
+        datakeys = set(datadict.keys())
+        reqkeys = {y for x in keypairs for y in x}
+        missing = not reqkeys.issubset(datakeys)
+
+        if missing:
+            errmsg = (
+                "Content validation failed.\n\
+                    Cannot compare missing keys: %s "
+                % (missing)
+            )
+            raise SchemaError(errmsg)
+
+        for k1, k2 in keypairs:
+            v1 = datadict[k1]
+            v2 = datadict[k2]
+            if v1 != v2:
+                errmsg = (
+                    "Content validation failed.\n\
+                          (d[%s] = %s) != (d[%s] = %s)"
+                    % (k1, v1, k2, v2)
+                )
+                raise SchemaError(errmsg)
+
+    @staticmethod
+    def load_csv_header(fname, dialect=None):
+        """Load the csv header.
+
+        Args:
+          fname(str): File name
+          dialect(dict): The csv dialect
+
+        Returns:
+          list: A list of strings representing the csv header.
+
+        Raises:
+          csv.Error: if invalid csv file.
+
+        """
+        if dialect is None:
+            dialect = DIALECT_CSV
+
+        with open(fname) as file:
+            reader = csv.reader(file, **dialect)
+            return next(reader)
+
+    @staticmethod
+    def load_csv(fname, dialect=None, bulk_load=True, skip_every=0):
+        """Load a CSV file.
+
+        This function can load the csv
+        file by bulk_loading or by reading row by row
+        with skip support.
+
+        Args:
+          fname(str): File name
+          dialect(dict): The csv dialect
+          bulk_load(bool): True for bulkload, False to read row by row
+          skip_every(int): read row interval
+
+        Returns:
+          list: row_numbers - list of row number read
+          dict: header - The header dict
+          list: data_rows - list with data rows
+
+        Raises:
+          csv.Error: if invalid csv file.
+
+        """
+        if dialect is None:
+            dialect = DIALECT_CSV
+
+        with open(fname) as file:
+            reader = csv.reader(file, **dialect)
+            header = next(reader)
+            if bulk_load:
+                data_rows = list(reader)
+                row_numbers = range(1, len(data_rows))
+                if skip_every:
+                    data_rows = data_rows[::skip_every]
+                    row_numbers = row_numbers[::skip_every]
+            else:
+                data_rows = []
+                row_numbers = []
+                for nrow, row in enumerate(reader):
+                    if skip_every:
+                        if nrow % skip_every:
+                            data_rows += [row]
+                            row_numbers += [nrow]
+                    else:
+                        data_rows += [row]
+                        row_numbers += [nrow]
+            return row_numbers, header, data_rows
```

Comment:
If you meant for reading csv's, Yes, we could used pandas. However, using it incur overhead of import (this class without it is much lighter), reading overhead (read_csv is slower than above), and API overhead (conversions from Dataframe/Series to list/dicts).

The main thing here is to fast load the data into lists (so set operations are quick) and column:data dicts. I would need to wait longer, convert frequently (Series to lists and double dictionary conversions because, surprisingly, pandas create index based dictionaries!!!!)

If you meant to leverage the "converters" argument of ```read_csv```, them the story is bit more complex and not beautiful either - I would need to wrap all schema calls to follow the pandas API expectation of that argument.